vyxal/parse.py

```diff
@@ -270,58 +153,87 @@ def parse(tokens: list[lexer.Token]) -> list[Structure]:
                 if len(branches) > 1:
-                    branches[-1] = parse(branches[-1])
-            elif structure_name == StructureType.FUNCTION_REF:
-                branches[0] = variable_name(branches)
-            elif structure_name == StructureType.LAMBDA:
+                    # It's got a body, so it's a function definition
+                    body = parse(branches[-1])
+                    structures.append(
+                        structure.FunctionDef(name, parameters, body)
+                    )
+                else:
+                    # No body, so it's a function call
+                    assert not parameters
+                    structures.append(structure.FunctionCall(name))
+
+            elif structure_cls == structure.FunctionReference:
+                structures.append(
+                    structure.FunctionReference(variable_name(branches[0]))
+                )
+
+            elif structure_cls == structure.Lambda:
                 if len(branches) == 1:
                     # that is, there is only a body - no arity
-                    branches.insert(0, "1")
+                    arity = 1
                 else:
-                    branches[0] = branches[0].value
-                branches[1] = parse(branches[1])
-            elif structure_name == StructureType.LAMBDA_MAP:
-                branches.insert(0, "1")
-                branches[1] = parse(branches[1])
-                structure_name = StructureType.LAMBDA
-                after_token = Structure(
-                    StructureType.NONE,
-                    lexer.Token(lexer.TokenType.GENERAL, "M"),
+                    try:
+                        arity = int(branches[0][0].value)
+                    except ValueError as ve:
+                        raise ValueError(
+                            "Arity must be parseable as an integer"
+                        ) from ve
+                    if arity < 0:
+                        raise ValueError("Arity must be non-negative")
+                structures.append(structure.Lambda(arity, parse(branches[-1])))
+
+            elif structure_cls == structure.LambdaMap:
+                structures.append(structure.Lambda(1, parse(branches[0])))
+                structures.append(
+                    structure.GenericStatement(
+                        [lexer.Token(lexer.TokenType.GENERAL, "M")]
+                    )
                 )
                 # laziness ftw
-            elif structure_name == StructureType.LAMBDA_FILTER:
-                branches.insert(0, "1")
-                branches[1] = parse(branches[1])
-                structure_name = StructureType.LAMBDA
-                after_token = Structure(
-                    StructureType.NONE,
-                    lexer.Token(lexer.TokenType.GENERAL, "F"),
+
+            elif structure_cls == structure.LambdaFilter:
+                structures.append(structure.Lambda(1, parse(branches[0])))
```

Comment:
What casing are we using?

Reply:
wdym?

Reply:
Heh, looks like I can reply from VSCode now.

Reply:
looks like CamelCase for classes and snake_case for variable, @chunkybanana.

Reply:
This is actually covered in the code guide: https://github.com/Vyxal/Vyxal/blob/fb-transpiler/documents/protocols/code%20guide.md